// @vue-flow/additional-components@1.3.3/dist/vue-flow-additional-components.mjs
import { defineComponent as zt, computed as I, openBlock as C, createElementBlock as B, normalizeStyle as he, renderSlot as H, createElementVNode as Z, unref as M, createCommentVNode as rt, normalizeClass as ne, toRef as at, createBlock as tt, withCtx as ht, Fragment as Fe, createVNode as Jt, resolveDynamicComponent as vt, inject as mn, h as gn, useAttrs as vn, ref as pn, provide as wn, useSlots as yn, watchEffect as _n, toDisplayString as xn, renderList as bn } from "vue";
import { useVueFlow as Pt, getRectOfNodes as kn, getBoundsofRects as $n, getConnectedEdges as pt } from "@vue-flow/core";
const Mn = ["x", "y", "width", "height"], zn = ["stroke", "stroke-width", "d"], Cn = ["cx", "cy", "r", "fill"], Nn = {
  key: 2,
  height: "100",
  width: "100"
}, En = ["fill"], Sn = ["x", "y", "fill"], An = {
  name: "Background"
}, Gs = /* @__PURE__ */ zt({
  ...An,
  props: {
    variant: { default: () => "dots" },
    gap: { default: 10 },
    size: { default: 0.4 },
    patternColor: null,
    bgColor: null,
    height: { default: 100 },
    width: { default: 100 },
    x: { default: 0 },
    y: { default: 0 }
  },
  setup(t) {
    const e = {
      dots: "#81818a",
      lines: "#eee"
    }, { id: n, viewport: i } = Pt(), r = I(() => {
      const u = t.gap && t.gap * i.value.zoom, l = u && i.value.x % u, f = u && i.value.y % u, g = t.size * i.value.zoom;
      return {
        scaledGap: u,
        xOffset: l,
        yOffset: f,
        size: g
      };
    }), o = `pattern-${n}`, s = I(() => t.patternColor || e[t.variant || "dots"]), a = I(
      () => `M${r.value.scaledGap / 2} 0 V${r.value.scaledGap} M0 ${r.value.scaledGap / 2} H${r.value.scaledGap}`
    );
    return (u, l) => (C(), B("svg", {
      class: "vue-flow__background",
      style: he({
        height: `${t.height > 100 ? 100 : t.height}%`,
        width: `${t.width > 100 ? 100 : t.width}%`
      })
    }, [
      H(u.$slots, "pattern-container", { id: o }, () => [
        Z("pattern", {
          id: o,
          x: r.value.xOffset,
          y: r.value.yOffset,
          width: r.value.scaledGap,
          height: r.value.scaledGap,
          patternUnits: "userSpaceOnUse"
        }, [
          H(u.$slots, "pattern", {}, () => [
            t.variant === "lines" ? (C(), B("path", {
              key: 0,
              stroke: M(s),
              "stroke-width": t.size,
              d: M(a)
            }, null, 8, zn)) : t.variant === "dots" ? (C(), B("circle", {
              key: 1,
              cx: r.value.size,
              cy: r.value.size,
              r: r.value.size,
              fill: M(s)
            }, null, 8, Cn)) : rt("", !0),
            t.bgColor ? (C(), B("svg", Nn, [
              Z("rect", {
                width: "100%",
                height: "100%",
                fill: t.bgColor
              }, null, 8, En)
            ])) : rt("", !0)
          ])
        ], 8, Mn)
      ]),
      Z("rect", {
        x: t.x,
        y: t.y,
        width: "100%",
        height: "100%",
        fill: `url(#${o})`
      }, null, 8, Sn),
      H(u.$slots, "default", { id: o })
    ], 4));
  }
});
var Tn = /* @__PURE__ */ ((t) => (t.Lines = "lines", t.Dots = "dots", t))(Tn || {});
const Oe = /* @__PURE__ */ zt({
  __name: "Panel",
  props: {
    position: null
  },
  setup(t) {
    const e = t, { userSelectionActive: n } = Pt(), i = I(() => `${e.position}`.split("-"));
    return (r, o) => (C(), B("div", {
      class: ne(["vue-flow__panel", M(i)]),
      style: he({ pointerEvents: M(n) ? "none" : "all" })
    }, [
      H(r.$slots, "default")
    ], 6));
  }
});
var Rn = /* @__PURE__ */ ((t) => (t.TopLeft = "top-left", t.TopCenter = "top-center", t.TopRight = "top-right", t.BottomLeft = "bottom-left", t.BottomCenter = "bottom-center", t.BottomRight = "bottom-right", t))(Rn || {});
const In = {
  name: "ControlButton"
}, Hn = (t, e) => {
  const n = t.__vccOpts || t;
  for (const [i, r] of e)
    n[i] = r;
  return n;
}, Vn = { class: "vue-flow__controls-button" };
function Dn(t, e, n, i, r, o) {
  return C(), B("button", Vn, [
    H(t.$slots, "default")
  ]);
}
const Et = /* @__PURE__ */ Hn(In, [["render", Dn]]), Bn = {
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 32 32"
}, Fn = /* @__PURE__ */ Z("path", { d: "M32 18.133H18.133V32h-4.266V18.133H0v-4.266h13.867V0h4.266v13.867H32z" }, null, -1), On = [
  Fn
];
function Xn(t, e) {
  return C(), B("svg", Bn, On);
}
const qn = { render: Xn }, Ln = {
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 32 5"
}, Pn = /* @__PURE__ */ Z("path", { d: "M0 0h32v4.2H0z" }, null, -1), Yn = [
  Pn
];
function Gn(t, e) {
  return C(), B("svg", Ln, Yn);
}
const Wn = { render: Gn }, Un = {
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 32 30"
}, Zn = /* @__PURE__ */ Z("path", { d: "M3.692 4.63c0-.53.4-.938.939-.938h5.215V0H4.708C2.13 0 0 2.054 0 4.63v5.216h3.692V4.631zM27.354 0h-5.2v3.692h5.17c.53 0 .984.4.984.939v5.215H32V4.631A4.624 4.624 0 0 0 27.354 0zm.954 24.83c0 .532-.4.94-.939.94h-5.215v3.768h5.215c2.577 0 4.631-2.13 4.631-4.707v-5.139h-3.692v5.139zm-23.677.94a.919.919 0 0 1-.939-.94v-5.138H0v5.139c0 2.577 2.13 4.707 4.708 4.707h5.138V25.77H4.631z" }, null, -1), Kn = [
  Zn
];
function Qn(t, e) {
  return C(), B("svg", Un, Kn);
}
const Jn = { render: Qn }, jn = {
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 25 32"
}, ti = /* @__PURE__ */ Z("path", { d: "M21.333 10.667H19.81V7.619C19.81 3.429 16.38 0 12.19 0 8 0 4.571 3.429 4.571 7.619v3.048H3.048A3.056 3.056 0 0 0 0 13.714v15.238A3.056 3.056 0 0 0 3.048 32h18.285a3.056 3.056 0 0 0 3.048-3.048V13.714a3.056 3.056 0 0 0-3.048-3.047zM12.19 24.533a3.056 3.056 0 0 1-3.047-3.047 3.056 3.056 0 0 1 3.047-3.048 3.056 3.056 0 0 1 3.048 3.048 3.056 3.056 0 0 1-3.048 3.047zm4.724-13.866H7.467V7.619c0-2.59 2.133-4.724 4.723-4.724 2.591 0 4.724 2.133 4.724 4.724v3.048z" }, null, -1), ei = [
  ti
];
function ni(t, e) {
  return C(), B("svg", jn, ei);
}
const ii = { render: ni }, ri = {
  xmlns: "http://www.w3.org/2000/svg",
  viewBox: "0 0 25 32"
}, oi = /* @__PURE__ */ Z("path", { d: "M21.333 10.667H19.81V7.619C19.81 3.429 16.38 0 12.19 0c-4.114 1.828-1.37 2.133.305 2.438 1.676.305 4.42 2.59 4.42 5.181v3.048H3.047A3.056 3.056 0 0 0 0 13.714v15.238A3.056 3.056 0 0 0 3.048 32h18.285a3.056 3.056 0 0 0 3.048-3.048V13.714a3.056 3.056 0 0 0-3.048-3.047zM12.19 24.533a3.056 3.056 0 0 1-3.047-3.047 3.056 3.056 0 0 1 3.047-3.048 3.056 3.056 0 0 1 3.048 3.048 3.056 3.056 0 0 1-3.048 3.047z" }, null, -1), si = [
  oi
];
function ai(t, e) {
  return C(), B("svg", ri, si);
}
const ui = { render: ai }, li = {
  name: "Controls"
}, Ws = /* @__PURE__ */ zt({
  ...li,
  props: {
    showZoom: { type: Boolean, default: !0 },
    showFitView: { type: Boolean, default: !0 },
    showInteractive: { type: Boolean, default: !0 },
    fitViewParams: null,
    position: { default: () => "bottom-left" }
  },
  emits: ["zoomIn", "zoomOut", "fitView", "interactionChange"],
  setup(t, { emit: e }) {
    const n = Pt(), i = at(n, "nodesDraggable"), r = at(n, "nodesConnectable"), o = at(n, "elementsSelectable"), s = at(n, "setInteractive"), a = at(n, "zoomIn"), u = at(n, "zoomOut"), l = at(n, "fitView"), f = I(() => i.value && r.value && o.value), g = () => {
      a.value(), e("zoomIn");
    }, v = () => {
      u.value(), e("zoomOut");
    }, w = () => {
      l.value(t.fitViewParams), e("fitView");
    }, k = () => {
      s.value(!f.value), e("interactionChange", !f.value);
    };
    return (b, N) => (C(), tt(M(Oe), {
      class: "vue-flow__controls",
      position: t.position
    }, {
      default: ht(() => [
        H(b.$slots, "top"),
        t.showZoom ? (C(), B(Fe, { key: 0 }, [
          H(b.$slots, "control-zoom-in", {}, () => [
            Jt(Et, {
              class: "vue-flow__controls-zoomin",
              onClick: g
            }, {
              default: ht(() => [
                H(b.$slots, "icon-zoom-in", {}, () => [
                  (C(), tt(vt(M(qn))))
                ])
              ]),
              _: 3
            })
          ]),
          H(b.$slots, "control-zoom-out", {}, () => [
            Jt(Et, {
              class: "vue-flow__controls-zoomout",
              onClick: v
            }, {
              default: ht(() => [
                H(b.$slots, "icon-zoom-out", {}, () => [
                  (C(), tt(vt(M(Wn))))
                ])
              ]),
              _: 3
            })
          ])
        ], 64)) : rt("", !0),
        t.showFitView ? H(b.$slots, "control-fit-view", { key: 1 }, () => [
          Jt(Et, {
            class: "vue-flow__controls-fitview",
            onClick: w
          }, {
            default: ht(() => [
              H(b.$slots, "icon-fit-view", {}, () => [
                (C(), tt(vt(M(Jn))))
              ])
            ]),
            _: 3
          })
        ]) : rt("", !0),
        t.showInteractive ? H(b.$slots, "control-interactive", { key: 2 }, () => [
          t.showInteractive ? (C(), tt(Et, {
            key: 0,
            class: "vue-flow__controls-interactive",
            onClick: k
          }, {
            default: ht(() => [
              M(f) ? H(b.$slots, "icon-unlock", { key: 0 }, () => [
                (C(), tt(vt(M(ui))))
              ]) : rt("", !0),
              M(f) ? rt("", !0) : H(b.$slots, "icon-lock", { key: 1 }, () => [
                (C(), tt(vt(M(ii))))
              ])
            ]),
            _: 3
          })) : rt("", !0)
        ]) : rt("", !0),
        H(b.$slots, "default")
      ]),
      _: 3
    }, 8, ["position"]));
  }
});
var ci = { value: () => {
} };
function de() {
  for (var t = 0, e = arguments.length, n = {}, i; t < e; ++t) {
    if (!(i = arguments[t] + "") || i in n || /[\s.]/.test(i))
      throw new Error("illegal type: " + i);
    n[i] = [];
  }
  return new It(n);
}
function It(t) {
  this._ = t;
}
function fi(t, e) {
  return t.trim().split(/^|\s+/).map(function(n) {
    var i = "", r = n.indexOf(".");
    if (r >= 0 && (i = n.slice(r + 1), n = n.slice(0, r)), n && !e.hasOwnProperty(n))
      throw new Error("unknown type: " + n);
    return { type: n, name: i };
  });
}
It.prototype = de.prototype = {
  constructor: It,
  on: function(t, e) {
    var n = this._, i = fi(t + "", n), r, o = -1, s = i.length;
    if (arguments.length < 2) {
      for (; ++o < s; )
        if ((r = (t = i[o]).type) && (r = hi(n[r], t.name)))
          return r;
      return;
    }
    if (e != null && typeof e != "function")
      throw new Error("invalid callback: " + e);
    for (; ++o < s; )
      if (r = (t = i[o]).type)
        n[r] = xe(n[r], t.name, e);
      else if (e == null)
        for (r in n)
          n[r] = xe(n[r], t.name, null);
    return this;
  },
  copy: function() {
    var t = {}, e = this._;
    for (var n in e)
      t[n] = e[n].slice();
    return new It(t);
  },
  call: function(t, e) {
    if ((r = arguments.length - 2) > 0)
      for (var n = new Array(r), i = 0, r, o; i < r; ++i)
        n[i] = arguments[i + 2];
    if (!this._.hasOwnProperty(t))
      throw new Error("unknown type: " + t);
    for (o = this._[t], i = 0, r = o.length; i < r; ++i)
      o[i].value.apply(e, n);
  },
  apply: function(t, e, n) {
    if (!this._.hasOwnProperty(t))
      throw new Error("unknown type: " + t);
    for (var i = this._[t], r = 0, o = i.length; r < o; ++r)
      i[r].value.apply(e, n);
  }
};
function hi(t, e) {
  for (var n = 0, i = t.length, r; n < i; ++n)
    if ((r = t[n]).name === e)
      return r.value;
}
function xe(t, e, n) {
  for (var i = 0, r = t.length; i < r; ++i)
    if (t[i].name === e) {
      t[i] = ci, t = t.slice(0, i).concat(t.slice(i + 1));
      break;
    }
  return n != null && t.push({ name: e, value: n }), t;
}
var ie = "http://www.w3.org/1999/xhtml";
const be = {
  svg: "http://www.w3.org/2000/svg",
  xhtml: ie,
  xlink: "http://www.w3.org/1999/xlink",
  xml: "http://www.w3.org/XML/1998/namespace",
  xmlns: "http://www.w3.org/2000/xmlns/"
};
function Yt(t) {
  var e = t += "", n = e.indexOf(":");
  return n >= 0 && (e = t.slice(0, n)) !== "xmlns" && (t = t.slice(n + 1)), be.hasOwnProperty(e) ? { space: be[e], local: t } : t;
}
function di(t) {
  return function() {
    var e = this.ownerDocument, n = this.namespaceURI;
    return n === ie && e.documentElement.namespaceURI === ie ? e.createElement(t) : e.createElementNS(n, t);
  };
}
function mi(t) {
  return function() {
    return this.ownerDocument.createElementNS(t.space, t.local);
  };
}
function Xe(t) {
  var e = Yt(t);
  return (e.local ? mi : di)(e);
}
function gi() {
}
function me(t) {
  return t == null ? gi : function() {
    return this.querySelector(t);
  };
}
function vi(t) {
  typeof t != "function" && (t = me(t));
  for (var e = this._groups, n = e.length, i = new Array(n), r = 0; r < n; ++r)
    for (var o = e[r], s = o.length, a = i[r] = new Array(s), u, l, f = 0; f < s; ++f)
      (u = o[f]) && (l = t.call(u, u.__data__, f, o)) && ("__data__" in u && (l.__data__ = u.__data__), a[f] = l);
  return new q(i, this._parents);
}
function pi(t) {
  return t == null ? [] : Array.isArray(t) ? t : Array.from(t);
}
function wi() {
  return [];
}
function qe(t) {
  return t == null ? wi : function() {
    return this.querySelectorAll(t);
  };
}
function yi(t) {
  return function() {
    return pi(t.apply(this, arguments));
  };
}
function _i(t) {
  typeof t == "function" ? t = yi(t) : t = qe(t);
  for (var e = this._groups, n = e.length, i = [], r = [], o = 0; o < n; ++o)
    for (var s = e[o], a = s.length, u, l = 0; l < a; ++l)
      (u = s[l]) && (i.push(t.call(u, u.__data__, l, s)), r.push(u));
  return new q(i, r);
}
function Le(t) {
  return function() {
    return this.matches(t);
  };
}
function Pe(t) {
  return function(e) {
    return e.matches(t);
  };
}
var xi = Array.prototype.find;
function bi(t) {
  return function() {
    return xi.call(this.children, t);
  };
}
function ki() {
  return this.firstElementChild;
}
function $i(t) {
  return this.select(t == null ? ki : bi(typeof t == "function" ? t : Pe(t)));
}
var Mi = Array.prototype.filter;
function zi() {
  return Array.from(this.children);
}
function Ci(t) {
  return function() {
    return Mi.call(this.children, t);
  };
}
function Ni(t) {
  return this.selectAll(t == null ? zi : Ci(typeof t == "function" ? t : Pe(t)));
}
function Ei(t) {
  typeof t != "function" && (t = Le(t));
  for (var e = this._groups, n = e.length, i = new Array(n), r = 0; r < n; ++r)
    for (var o = e[r], s = o.length, a = i[r] = [], u, l = 0; l < s; ++l)
      (u = o[l]) && t.call(u, u.__data__, l, o) && a.push(u);
  return new q(i, this._parents);
}
function Ye(t) {
  return new Array(t.length);
}
function Si() {
  return new q(this._enter || this._groups.map(Ye), this._parents);
}
function Bt(t, e) {
  this.ownerDocument = t.ownerDocument, this.namespaceURI = t.namespaceURI, this._next = null, this._parent = t, this.__data__ = e;
}
Bt.prototype = {
  constructor: Bt,
  appendChild: function(t) {
    return this._parent.insertBefore(t, this._next);
  },
  insertBefore: function(t, e) {
    return this._parent.insertBefore(t, e);
  },
  querySelector: function(t) {
    return this._parent.querySelector(t);
  },
  querySelectorAll: function(t) {
    return this._parent.querySelectorAll(t);
  }
};
function Ai(t) {
  return function() {
    return t;
  };
}
function Ti(t, e, n, i, r, o) {
  for (var s = 0, a, u = e.length, l = o.length; s < l; ++s)
    (a = e[s]) ? (a.__data__ = o[s], i[s] = a) : n[s] = new Bt(t, o[s]);
  for (; s < u; ++s)
    (a = e[s]) && (r[s] = a);
}
function Ri(t, e, n, i, r, o, s) {
  var a, u, l = /* @__PURE__ */ new Map(), f = e.length, g = o.length, v = new Array(f), w;
  for (a = 0; a < f; ++a)
    (u = e[a]) && (v[a] = w = s.call(u, u.__data__, a, e) + "", l.has(w) ? r[a] = u : l.set(w, u));
  for (a = 0; a < g; ++a)
    w = s.call(t, o[a], a, o) + "", (u = l.get(w)) ? (i[a] = u, u.__data__ = o[a], l.delete(w)) : n[a] = new Bt(t, o[a]);
  for (a = 0; a < f; ++a)
    (u = e[a]) && l.get(v[a]) === u && (r[a] = u);
}
function Ii(t) {
  return t.__data__;
}
function Hi(t, e) {
  if (!arguments.length)
    return Array.from(this, Ii);
  var n = e ? Ri : Ti, i = this._parents, r = this._groups;
  typeof t != "function" && (t = Ai(t));
  for (var o = r.length, s = new Array(o), a = new Array(o), u = new Array(o), l = 0; l < o; ++l) {
    var f = i[l], g = r[l], v = g.length, w = Vi(t.call(f, f && f.__data__, l, i)), k = w.length, b = a[l] = new Array(k), N = s[l] = new Array(k), x = u[l] = new Array(v);
    n(f, g, b, N, x, w, e);
    for (var T = 0, R = 0, L, O; T < k; ++T)
      if (L = b[T]) {
        for (T >= R && (R = T + 1); !(O = N[R]) && ++R < k; )
          ;
        L._next = O || null;
      }
  }
  return s = new q(s, i), s._enter = a, s._exit = u, s;
}
function Vi(t) {
  return typeof t == "object" && "length" in t ? t : Array.from(t);
}
function Di() {
  return new q(this._exit || this._groups.map(Ye), this._parents);
}
function Bi(t, e, n) {
  var i = this.enter(), r = this, o = this.exit();
  return typeof t == "function" ? (i = t(i), i && (i = i.selection())) : i = i.append(t + ""), e != null && (r = e(r), r && (r = r.selection())), n == null ? o.remove() : n(o), i && r ? i.merge(r).order() : r;
}
function Fi(t) {
  for (var e = t.selection ? t.selection() : t, n = this._groups, i = e._groups, r = n.length, o = i.length, s = Math.min(r, o), a = new Array(r), u = 0; u < s; ++u)
    for (var l = n[u], f = i[u], g = l.length, v = a[u] = new Array(g), w, k = 0; k < g; ++k)
      (w = l[k] || f[k]) && (v[k] = w);
  for (; u < r; ++u)
    a[u] = n[u];
  return new q(a, this._parents);
}
function Oi() {
  for (var t = this._groups, e = -1, n = t.length; ++e < n; )
    for (var i = t[e], r = i.length - 1, o = i[r], s; --r >= 0; )
      (s = i[r]) && (o && s.compareDocumentPosition(o) ^ 4 && o.parentNode.insertBefore(s, o), o = s);
  return this;
}
function Xi(t) {
  t || (t = qi);
  function e(g, v) {
    return g && v ? t(g.__data__, v.__data__) : !g - !v;
  }
  for (var n = this._groups, i = n.length, r = new Array(i), o = 0; o < i; ++o) {
    for (var s = n[o], a = s.length, u = r[o] = new Array(a), l, f = 0; f < a; ++f)
      (l = s[f]) && (u[f] = l);
    u.sort(e);
  }
  return new q(r, this._parents).order();
}
function qi(t, e) {
  return t < e ? -1 : t > e ? 1 : t >= e ? 0 : NaN;
}
function Li() {
  var t = arguments[0];
  return arguments[0] = this, t.apply(null, arguments), this;
}
function Pi() {
  return Array.from(this);
}
function Yi() {
  for (var t = this._groups, e = 0, n = t.length; e < n; ++e)
    for (var i = t[e], r = 0, o = i.length; r < o; ++r) {
      var s = i[r];
      if (s)
        return s;
    }
  return null;
}
function Gi() {
  let t = 0;
  for (const e of this)
    ++t;
  return t;
}
function Wi() {
  return !this.node();
}
function Ui(t) {
  for (var e = this._groups, n = 0, i = e.length; n < i; ++n)
    for (var r = e[n], o = 0, s = r.length, a; o < s; ++o)
      (a = r[o]) && t.call(a, a.__data__, o, r);
  return this;
}
function Zi(t) {
  return function() {
    this.removeAttribute(t);
  };
}
function Ki(t) {
  return function() {
    this.removeAttributeNS(t.space, t.local);
  };
}
function Qi(t, e) {
  return function() {
    this.setAttribute(t, e);
  };
}
function Ji(t, e) {
  return function() {
    this.setAttributeNS(t.space, t.local, e);
  };
}
function ji(t, e) {
  return function() {
    var n = e.apply(this, arguments);
    n == null ? this.removeAttribute(t) : this.setAttribute(t, n);
  };
}
function tr(t, e) {
  return function() {
    var n = e.apply(this, arguments);
    n == null ? this.removeAttributeNS(t.space, t.local) : this.setAttributeNS(t.space, t.local, n);
  };
}
function er(t, e) {
  var n = Yt(t);
  if (arguments.length < 2) {
    var i = this.node();
    return n.local ? i.getAttributeNS(n.space, n.local) : i.getAttribute(n);
  }
  return this.each((e == null ? n.local ? Ki : Zi : typeof e == "function" ? n.local ? tr : ji : n.local ? Ji : Qi)(n, e));
}
function Ge(t) {
  return t.ownerDocument && t.ownerDocument.defaultView || t.document && t || t.defaultView;
}
function nr(t) {
  return function() {
    this.style.removeProperty(t);
  };
}
function ir(t, e, n) {
  return function() {
    this.style.setProperty(t, e, n);
  };
}
function rr(t, e, n) {
  return function() {
    var i = e.apply(this, arguments);
    i == null ? this.style.removeProperty(t) : this.style.setProperty(t, i, n);
  };
}
function or(t, e, n) {
  return arguments.length > 1 ? this.each((e == null ? nr : typeof e == "function" ? rr : ir)(t, e, n == null ? "" : n)) : mt(this.node(), t);
}
function mt(t, e) {
  return t.style.getPropertyValue(e) || Ge(t).getComputedStyle(t, null).getPropertyValue(e);
}
function sr(t) {
  return function() {
    delete this[t];
  };
}
function ar(t, e) {
  return function() {
    this[t] = e;
  };
}
function ur(t, e) {
  return function() {
    var n = e.apply(this, arguments);
    n == null ? delete this[t] : this[t] = n;
  };
}
function lr(t, e) {
  return arguments.length > 1 ? this.each((e == null ? sr : typeof e == "function" ? ur : ar)(t, e)) : this.node()[t];
}
function We(t) {
  return t.trim().split(/^|\s+/);
}
function ge(t) {
  return t.classList || new Ue(t);
}
function Ue(t) {
  this._node = t, this._names = We(t.getAttribute("class") || "");
}
Ue.prototype = {
  add: function(t) {
    var e = this._names.indexOf(t);
    e < 0 && (this._names.push(t), this._node.setAttribute("class", this._names.join(" ")));
  },
  remove: function(t) {
    var e = this._names.indexOf(t);
    e >= 0 && (this._names.splice(e, 1), this._node.setAttribute("class", this._names.join(" ")));
  },
  contains: function(t) {
    return this._names.indexOf(t) >= 0;
  }
};
function Ze(t, e) {
  for (var n = ge(t), i = -1, r = e.length; ++i < r; )
    n.add(e[i]);
}
function Ke(t, e) {
  for (var n = ge(t), i = -1, r = e.length; ++i < r; )
    n.remove(e[i]);
}
function cr(t) {
  return function() {
    Ze(this, t);
  };
}
function fr(t) {
  return function() {
    Ke(this, t);
  };
}
function hr(t, e) {
  return function() {
    (e.apply(this, arguments) ? Ze : Ke)(this, t);
  };
}
function dr(t, e) {
  var n = We(t + "");
  if (arguments.length < 2) {
    for (var i = ge(this.node()), r = -1, o = n.length; ++r < o; )
      if (!i.contains(n[r]))
        return !1;
    return !0;
  }
  return this.each((typeof e == "function" ? hr : e ? cr : fr)(n, e));
}
function mr() {
  this.textContent = "";
}
function gr(t) {
  return function() {
    this.textContent = t;
  };
}
function vr(t) {
  return function() {
    var e = t.apply(this, arguments);
    this.textContent = e == null ? "" : e;
  };
}
function pr(t) {
  return arguments.length ? this.each(t == null ? mr : (typeof t == "function" ? vr : gr)(t)) : this.node().textContent;
}
function wr() {
  this.innerHTML = "";
}
function yr(t) {
  return function() {
    this.innerHTML = t;
  };
}
function _r(t) {
  return function() {
    var e = t.apply(this, arguments);
    this.innerHTML = e == null ? "" : e;
  };
}
function xr(t) {
  return arguments.length ? this.each(t == null ? wr : (typeof t == "function" ? _r : yr)(t)) : this.node().innerHTML;
}
function br() {
  this.nextSibling && this.parentNode.appendChild(this);
}
function kr() {
  return this.each(br);
}
function $r() {
  this.previousSibling && this.parentNode.insertBefore(this, this.parentNode.firstChild);
}
function Mr() {
  return this.each($r);
}
function zr(t) {
  var e = typeof t == "function" ? t : Xe(t);
  return this.select(function() {
    return this.appendChild(e.apply(this, arguments));
  });
}
function Cr() {
  return null;
}
function Nr(t, e) {
  var n = typeof t == "function" ? t : Xe(t), i = e == null ? Cr : typeof e == "function" ? e : me(e);
  return this.select(function() {
    return this.insertBefore(n.apply(this, arguments), i.apply(this, arguments) || null);
  });
}
function Er() {
  var t = this.parentNode;
  t && t.removeChild(this);
}
function Sr() {
  return this.each(Er);
}
function Ar() {
  var t = this.cloneNode(!1), e = this.parentNode;
  return e ? e.insertBefore(t, this.nextSibling) : t;
}
function Tr() {
  var t = this.cloneNode(!0), e = this.parentNode;
  return e ? e.insertBefore(t, this.nextSibling) : t;
}
function Rr(t) {
  return this.select(t ? Tr : Ar);
}
function Ir(t) {
  return arguments.length ? this.property("__data__", t) : this.node().__data__;
}
function Hr(t) {
  return function(e) {
    t.call(this, e, this.__data__);
  };
}
function Vr(t) {
  return t.trim().split(/^|\s+/).map(function(e) {
    var n = "", i = e.indexOf(".");
    return i >= 0 && (n = e.slice(i + 1), e = e.slice(0, i)), { type: e, name: n };
  });
}
function Dr(t) {
  return function() {
    var e = this.__on;
    if (!!e) {
      for (var n = 0, i = -1, r = e.length, o; n < r; ++n)
        o = e[n], (!t.type || o.type === t.type) && o.name === t.name ? this.removeEventListener(o.type, o.listener, o.options) : e[++i] = o;
      ++i ? e.length = i : delete this.__on;
    }
  };
}
function Br(t, e, n) {
  return function() {
    var i = this.__on, r, o = Hr(e);
    if (i) {
      for (var s = 0, a = i.length; s < a; ++s)
        if ((r = i[s]).type === t.type && r.name === t.name) {
          this.removeEventListener(r.type, r.listener, r.options), this.addEventListener(r.type, r.listener = o, r.options = n), r.value = e;
          return;
        }
    }
    this.addEventListener(t.type, o, n), r = { type: t.type, name: t.name, value: e, listener: o, options: n }, i ? i.push(r) : this.__on = [r];
  };
}
function Fr(t, e, n) {
  var i = Vr(t + ""), r, o = i.length, s;
  if (arguments.length < 2) {
    var a = this.node().__on;
    if (a) {
      for (var u = 0, l = a.length, f; u < l; ++u)
        for (r = 0, f = a[u]; r < o; ++r)
          if ((s = i[r]).type === f.type && s.name === f.name)
            return f.value;
    }
    return;
  }
  for (a = e ? Br : Dr, r = 0; r < o; ++r)
    this.each(a(i[r], e, n));
  return this;
}
function Qe(t, e, n) {
  var i = Ge(t), r = i.CustomEvent;
  typeof r == "function" ? r = new r(e, n) : (r = i.document.createEvent("Event"), n ? (r.initEvent(e, n.bubbles, n.cancelable), r.detail = n.detail) : r.initEvent(e, !1, !1)), t.dispatchEvent(r);
}
function Or(t, e) {
  return function() {
    return Qe(this, t, e);
  };
}
function Xr(t, e) {
  return function() {
    return Qe(this, t, e.apply(this, arguments));
  };
}
function qr(t, e) {
  return this.each((typeof e == "function" ? Xr : Or)(t, e));
}
function* Lr() {
  for (var t = this._groups, e = 0, n = t.length; e < n; ++e)
    for (var i = t[e], r = 0, o = i.length, s; r < o; ++r)
      (s = i[r]) && (yield s);
}
var Je = [null];
function q(t, e) {
  this._groups = t, this._parents = e;
}
function Ct() {
  return new q([[document.documentElement]], Je);
}
function Pr() {
  return this;
}
q.prototype = Ct.prototype = {
  constructor: q,
  select: vi,
  selectAll: _i,
  selectChild: $i,
  selectChildren: Ni,
  filter: Ei,
  data: Hi,
  enter: Si,
  exit: Di,
  join: Bi,
  merge: Fi,
  selection: Pr,
  order: Oi,
  sort: Xi,
  call: Li,
  nodes: Pi,
  node: Yi,
  size: Gi,
  empty: Wi,
  each: Ui,
  attr: er,
  style: or,
  property: lr,
  classed: dr,
  text: pr,
  html: xr,
  raise: kr,
  lower: Mr,
  append: zr,
  insert: Nr,
  remove: Sr,
  clone: Rr,
  datum: Ir,
  on: Fr,
  dispatch: qr,
  [Symbol.iterator]: Lr
};
function st(t) {
  return typeof t == "string" ? new q([[document.querySelector(t)]], [document.documentElement]) : new q([[t]], Je);
}
function Yr(t) {
  let e;
  for (; e = t.sourceEvent; )
    t = e;
  return t;
}
function it(t, e) {
  if (t = Yr(t), e === void 0 && (e = t.currentTarget), e) {
    var n = e.ownerSVGElement || e;
    if (n.createSVGPoint) {
      var i = n.createSVGPoint();
      return i.x = t.clientX, i.y = t.clientY, i = i.matrixTransform(e.getScreenCTM().inverse()), [i.x, i.y];
    }
    if (e.getBoundingClientRect) {
      var r = e.getBoundingClientRect();
      return [t.clientX - r.left - e.clientLeft, t.clientY - r.top - e.clientTop];
    }
  }
  return [t.pageX, t.pageY];
}
const re = { capture: !0, passive: !1 };
function oe(t) {
  t.preventDefault(), t.stopImmediatePropagation();
}
function Gr(t) {
  var e = t.document.documentElement, n = st(t).on("dragstart.drag", oe, re);
  "onselectstart" in e ? n.on("selectstart.drag", oe, re) : (e.__noselect = e.style.MozUserSelect, e.style.MozUserSelect = "none");
}
function Wr(t, e) {
  var n = t.document.documentElement, i = st(t).on("dragstart.drag", null);
  e && (i.on("click.drag", oe, re), setTimeout(function() {
    i.on("click.drag", null);
  }, 0)), "onselectstart" in n ? i.on("selectstart.drag", null) : (n.style.MozUserSelect = n.__noselect, delete n.__noselect);
}
function ve(t, e, n) {
  t.prototype = e.prototype = n, n.constructor = t;
}
function je(t, e) {
  var n = Object.create(t.prototype);
  for (var i in e)
    n[i] = e[i];
  return n;
}
function Nt() {
}
var bt = 0.7, Ft = 1 / bt, dt = "\\s*([+-]?\\d+)\\s*", kt = "\\s*([+-]?(?:\\d*\\.)?\\d+(?:[eE][+-]?\\d+)?)\\s*", K = "\\s*([+-]?(?:\\d*\\.)?\\d+(?:[eE][+-]?\\d+)?)%\\s*", Ur = /^#([0-9a-f]{3,8})$/, Zr = new RegExp(`^rgb\\(${dt},${dt},${dt}\\)$`), Kr = new RegExp(`^rgb\\(${K},${K},${K}\\)$`), Qr = new RegExp(`^rgba\\(${dt},${dt},${dt},${kt}\\)$`), Jr = new RegExp(`^rgba\\(${K},${K},${K},${kt}\\)$`), jr = new RegExp(`^hsl\\(${kt},${K},${K}\\)$`), to = new RegExp(`^hsla\\(${kt},${K},${K},${kt}\\)$`), ke = {
  aliceblue: 15792383,
  antiquewhite: 16444375,
  aqua: 65535,
  aquamarine: 8388564,
  azure: 15794175,
  beige: 16119260,
  bisque: 16770244,
  black: 0,
  blanchedalmond: 16772045,
  blue: 255,
  blueviolet: 9055202,
  brown: 10824234,
  burlywood: 14596231,
  cadetblue: 6266528,
  chartreuse: 8388352,
  chocolate: 13789470,
  coral: 16744272,
  cornflowerblue: 6591981,
  cornsilk: 16775388,
  crimson: 14423100,
  cyan: 65535,
  darkblue: 139,
  darkcyan: 35723,
  darkgoldenrod: 12092939,
  darkgray: 11119017,
  darkgreen: 25600,
  darkgrey: 11119017,
  darkkhaki: 12433259,
  darkmagenta: 9109643,
  darkolivegreen: 5597999,
  darkorange: 16747520,
  darkorchid: 10040012,
  darkred: 9109504,
  darksalmon: 15308410,
  darkseagreen: 9419919,
  darkslateblue: 4734347,
  darkslategray: 3100495,
  darkslategrey: 3100495,
  darkturquoise: 52945,
  darkviolet: 9699539,
  deeppink: 16716947,
  deepskyblue: 49151,
  dimgray: 6908265,
  dimgrey: 6908265,
  dodgerblue: 2003199,
  firebrick: 11674146,
  floralwhite: 16775920,
  forestgreen: 2263842,
  fuchsia: 16711935,
  gainsboro: 14474460,
  ghostwhite: 16316671,
  gold: 16766720,
  goldenrod: 14329120,
  gray: 8421504,
  green: 32768,
  greenyellow: 11403055,
  grey: 8421504,
  honeydew: 15794160,
  hotpink: 16738740,
  indianred: 13458524,
  indigo: 4915330,
  ivory: 16777200,
  khaki: 15787660,
  lavender: 15132410,
  lavenderblush: 16773365,
  lawngreen: 8190976,
  lemonchiffon: 16775885,
  lightblue: 11393254,
  lightcoral: 15761536,
  lightcyan: 14745599,
  lightgoldenrodyellow: 16448210,
  lightgray: 13882323,
  lightgreen: 9498256,
  lightgrey: 13882323,
  lightpink: 16758465,
  lightsalmon: 16752762,
  lightseagreen: 2142890,
  lightskyblue: 8900346,
  lightslategray: 7833753,
  lightslategrey: 7833753,
  lightsteelblue: 11584734,
  lightyellow: 16777184,
  lime: 65280,
  limegreen: 3329330,
  linen: 16445670,
  magenta: 16711935,
  maroon: 8388608,
  mediumaquamarine: 6737322,
  mediumblue: 205,
  mediumorchid: 12211667,
  mediumpurple: 9662683,
  mediumseagreen: 3978097,
  mediumslateblue: 8087790,
  mediumspringgreen: 64154,
  mediumturquoise: 4772300,
  mediumvioletred: 13047173,
  midnightblue: 1644912,
  mintcream: 16121850,
  mistyrose: 16770273,
  moccasin: 16770229,
  navajowhite: 16768685,
  navy: 128,
  oldlace: 16643558,
  olive: 8421376,
  olivedrab: 7048739,
  orange: 16753920,
  orangered: 16729344,
  orchid: 14315734,
  palegoldenrod: 15657130,
  palegreen: 10025880,
  paleturquoise: 11529966,
  palevioletred: 14381203,
  papayawhip: 16773077,
  peachpuff: 16767673,
  peru: 13468991,
  pink: 16761035,
  plum: 14524637,
  powderblue: 11591910,
  purple: 8388736,
  rebeccapurple: 6697881,
  red: 16711680,
  rosybrown: 12357519,
  royalblue: 4286945,
  saddlebrown: 9127187,
  salmon: 16416882,
  sandybrown: 16032864,
  seagreen: 3050327,
  seashell: 16774638,
  sienna: 10506797,
  silver: 12632256,
  skyblue: 8900331,
  slateblue: 6970061,
  slategray: 7372944,
  slategrey: 7372944,
  snow: 16775930,
  springgreen: 65407,
  steelblue: 4620980,
  tan: 13808780,
  teal: 32896,
  thistle: 14204888,
  tomato: 16737095,
  turquoise: 4251856,
  violet: 15631086,
  wheat: 16113331,
  white: 16777215,
  whitesmoke: 16119285,
  yellow: 16776960,
  yellowgreen: 10145074
};
ve(Nt, $t, {
  copy(t) {
    return Object.assign(new this.constructor(), this, t);
  },
  displayable() {
    return this.rgb().displayable();
  },
  hex: $e,
  formatHex: $e,
  formatHex8: eo,
  formatHsl: no,
  formatRgb: Me,
  toString: Me
});
function $e() {
  return this.rgb().formatHex();
}
function eo() {
  return this.rgb().formatHex8();
}
function no() {
  return tn(this).formatHsl();
}
function Me() {
  return this.rgb().formatRgb();
}
function $t(t) {
  var e, n;
  return t = (t + "").trim().toLowerCase(), (e = Ur.exec(t)) ? (n = e[1].length, e = parseInt(e[1], 16), n === 6 ? ze(e) : n === 3 ? new F(e >> 8 & 15 | e >> 4 & 240, e >> 4 & 15 | e & 240, (e & 15) << 4 | e & 15, 1) : n === 8 ? St(e >> 24 & 255, e >> 16 & 255, e >> 8 & 255, (e & 255) / 255) : n === 4 ? St(e >> 12 & 15 | e >> 8 & 240, e >> 8 & 15 | e >> 4 & 240, e >> 4 & 15 | e & 240, ((e & 15) << 4 | e & 15) / 255) : null) : (e = Zr.exec(t)) ? new F(e[1], e[2], e[3], 1) : (e = Kr.exec(t)) ? new F(e[1] * 255 / 100, e[2] * 255 / 100, e[3] * 255 / 100, 1) : (e = Qr.exec(t)) ? St(e[1], e[2], e[3], e[4]) : (e = Jr.exec(t)) ? St(e[1] * 255 / 100, e[2] * 255 / 100, e[3] * 255 / 100, e[4]) : (e = jr.exec(t)) ? Ee(e[1], e[2] / 100, e[3] / 100, 1) : (e = to.exec(t)) ? Ee(e[1], e[2] / 100, e[3] / 100, e[4]) : ke.hasOwnProperty(t) ? ze(ke[t]) : t === "transparent" ? new F(NaN, NaN, NaN, 0) : null;
}
function ze(t) {
  return new F(t >> 16 & 255, t >> 8 & 255, t & 255, 1);
}
function St(t, e, n, i) {
  return i <= 0 && (t = e = n = NaN), new F(t, e, n, i);
}
function io(t) {
  return t instanceof Nt || (t = $t(t)), t ? (t = t.rgb(), new F(t.r, t.g, t.b, t.opacity)) : new F();
}
function se(t, e, n, i) {
  return arguments.length === 1 ? io(t) : new F(t, e, n, i == null ? 1 : i);
}
function F(t, e, n, i) {
  this.r = +t, this.g = +e, this.b = +n, this.opacity = +i;
}
ve(F, se, je(Nt, {
  brighter(t) {
    return t = t == null ? Ft : Math.pow(Ft, t), new F(this.r * t, this.g * t, this.b * t, this.opacity);
  },
  darker(t) {
    return t = t == null ? bt : Math.pow(bt, t), new F(this.r * t, this.g * t, this.b * t, this.opacity);
  },
  rgb() {
    return this;
  },
  clamp() {
    return new F(lt(this.r), lt(this.g), lt(this.b), Ot(this.opacity));
  },
  displayable() {
    return -0.5 <= this.r && this.r < 255.5 && -0.5 <= this.g && this.g < 255.5 && -0.5 <= this.b && this.b < 255.5 && 0 <= this.opacity && this.opacity <= 1;
  },
  hex: Ce,
  formatHex: Ce,
  formatHex8: ro,
  formatRgb: Ne,
  toString: Ne
}));
function Ce() {
  return `#${ut(this.r)}${ut(this.g)}${ut(this.b)}`;
}
function ro() {
  return `#${ut(this.r)}${ut(this.g)}${ut(this.b)}${ut((isNaN(this.opacity) ? 1 : this.opacity) * 255)}`;
}
function Ne() {
  const t = Ot(this.opacity);
  return `${t === 1 ? "rgb(" : "rgba("}${lt(this.r)}, ${lt(this.g)}, ${lt(this.b)}${t === 1 ? ")" : `, ${t})`}`;
}
function Ot(t) {
  return isNaN(t) ? 1 : Math.max(0, Math.min(1, t));
}
function lt(t) {
  return Math.max(0, Math.min(255, Math.round(t) || 0));
}
function ut(t) {
  return t = lt(t), (t < 16 ? "0" : "") + t.toString(16);
}
function Ee(t, e, n, i) {
  return i <= 0 ? t = e = n = NaN : n <= 0 || n >= 1 ? t = e = NaN : e <= 0 && (t = NaN), new G(t, e, n, i);
}
function tn(t) {
  if (t instanceof G)
    return new G(t.h, t.s, t.l, t.opacity);
  if (t instanceof Nt || (t = $t(t)), !t)
    return new G();
  if (t instanceof G)
    return t;
  t = t.rgb();
  var e = t.r / 255, n = t.g / 255, i = t.b / 255, r = Math.min(e, n, i), o = Math.max(e, n, i), s = NaN, a = o - r, u = (o + r) / 2;
  return a ? (e === o ? s = (n - i) / a + (n < i) * 6 : n === o ? s = (i - e) / a + 2 : s = (e - n) / a + 4, a /= u < 0.5 ? o + r : 2 - o - r, s *= 60) : a = u > 0 && u < 1 ? 0 : s, new G(s, a, u, t.opacity);
}
function oo(t, e, n, i) {
  return arguments.length === 1 ? tn(t) : new G(t, e, n, i == null ? 1 : i);
}
function G(t, e, n, i) {
  this.h = +t, this.s = +e, this.l = +n, this.opacity = +i;
}
ve(G, oo, je(Nt, {
  brighter(t) {
    return t = t == null ? Ft : Math.pow(Ft, t), new G(this.h, this.s, this.l * t, this.opacity);
  },
  darker(t) {
    return t = t == null ? bt : Math.pow(bt, t), new G(this.h, this.s, this.l * t, this.opacity);
  },
  rgb() {
    var t = this.h % 360 + (this.h < 0) * 360, e = isNaN(t) || isNaN(this.s) ? 0 : this.s, n = this.l, i = n + (n < 0.5 ? n : 1 - n) * e, r = 2 * n - i;
    return new F(
      jt(t >= 240 ? t - 240 : t + 120, r, i),
      jt(t, r, i),
      jt(t < 120 ? t + 240 : t - 120, r, i),
      this.opacity
    );
  },
  clamp() {
    return new G(Se(this.h), At(this.s), At(this.l), Ot(this.opacity));
  },
  displayable() {
    return (0 <= this.s && this.s <= 1 || isNaN(this.s)) && 0 <= this.l && this.l <= 1 && 0 <= this.opacity && this.opacity <= 1;
  },
  formatHsl() {
    const t = Ot(this.opacity);
    return `${t === 1 ? "hsl(" : "hsla("}${Se(this.h)}, ${At(this.s) * 100}%, ${At(this.l) * 100}%${t === 1 ? ")" : `, ${t})`}`;
  }
}));
function Se(t) {
  return t = (t || 0) % 360, t < 0 ? t + 360 : t;
}
function At(t) {
  return Math.max(0, Math.min(1, t || 0));
}
function jt(t, e, n) {
  return (t < 60 ? e + (n - e) * t / 60 : t < 180 ? n : t < 240 ? e + (n - e) * (240 - t) / 60 : e) * 255;
}
const en = (t) => () => t;
function so(t, e) {
  return function(n) {
    return t + n * e;
  };
}
function ao(t, e, n) {
  return t = Math.pow(t, n), e = Math.pow(e, n) - t, n = 1 / n, function(i) {
    return Math.pow(t + i * e, n);
  };
}
function uo(t) {
  return (t = +t) == 1 ? nn : function(e, n) {
    return n - e ? ao(e, n, t) : en(isNaN(e) ? n : e);
  };
}
function nn(t, e) {
  var n = e - t;
  return n ? so(t, n) : en(isNaN(t) ? e : t);
}
const Ae = function t(e) {
  var n = uo(e);
  function i(r, o) {
    var s = n((r = se(r)).r, (o = se(o)).r), a = n(r.g, o.g), u = n(r.b, o.b), l = nn(r.opacity, o.opacity);
    return function(f) {
      return r.r = s(f), r.g = a(f), r.b = u(f), r.opacity = l(f), r + "";
    };
  }
  return i.gamma = t, i;
}(1);
function ot(t, e) {
  return t = +t, e = +e, function(n) {
    return t * (1 - n) + e * n;
  };
}
var ae = /[-+]?(?:\d+\.?\d*|\.?\d+)(?:[eE][-+]?\d+)?/g, te = new RegExp(ae.source, "g");
function lo(t) {
  return function() {
    return t;
  };
}
function co(t) {
  return function(e) {
    return t(e) + "";
  };
}
function fo(t, e) {
  var n = ae.lastIndex = te.lastIndex = 0, i, r, o, s = -1, a = [], u = [];
  for (t = t + "", e = e + ""; (i = ae.exec(t)) && (r = te.exec(e)); )
    (o = r.index) > n && (o = e.slice(n, o), a[s] ? a[s] += o : a[++s] = o), (i = i[0]) === (r = r[0]) ? a[s] ? a[s] += r : a[++s] = r : (a[++s] = null, u.push({ i: s, x: ot(i, r) })), n = te.lastIndex;
  return n < e.length && (o = e.slice(n), a[s] ? a[s] += o : a[++s] = o), a.length < 2 ? u[0] ? co(u[0].x) : lo(e) : (e = u.length, function(l) {
    for (var f = 0, g; f < e; ++f)
      a[(g = u[f]).i] = g.x(l);
    return a.join("");
  });
}
var Te = 180 / Math.PI, ue = {
  translateX: 0,
  translateY: 0,
  rotate: 0,
  skewX: 0,
  scaleX: 1,
  scaleY: 1
};
function rn(t, e, n, i, r, o) {
  var s, a, u;
  return (s = Math.sqrt(t * t + e * e)) && (t /= s, e /= s), (u = t * n + e * i) && (n -= t * u, i -= e * u), (a = Math.sqrt(n * n + i * i)) && (n /= a, i /= a, u /= a), t * i < e * n && (t = -t, e = -e, u = -u, s = -s), {
    translateX: r,
    translateY: o,
    rotate: Math.atan2(e, t) * Te,
    skewX: Math.atan(u) * Te,
    scaleX: s,
    scaleY: a
  };
}
var Tt;
function ho(t) {
  const e = new (typeof DOMMatrix == "function" ? DOMMatrix : WebKitCSSMatrix)(t + "");
  return e.isIdentity ? ue : rn(e.a, e.b, e.c, e.d, e.e, e.f);
}
function mo(t) {
  return t == null || (Tt || (Tt = document.createElementNS("http://www.w3.org/2000/svg", "g")), Tt.setAttribute("transform", t), !(t = Tt.transform.baseVal.consolidate())) ? ue : (t = t.matrix, rn(t.a, t.b, t.c, t.d, t.e, t.f));
}
function on(t, e, n, i) {
  function r(l) {
    return l.length ? l.pop() + " " : "";
  }
  function o(l, f, g, v, w, k) {
    if (l !== g || f !== v) {
      var b = w.push("translate(", null, e, null, n);
      k.push({ i: b - 4, x: ot(l, g) }, { i: b - 2, x: ot(f, v) });
    } else
      (g || v) && w.push("translate(" + g + e + v + n);
  }
  function s(l, f, g, v) {
    l !== f ? (l - f > 180 ? f += 360 : f - l > 180 && (l += 360), v.push({ i: g.push(r(g) + "rotate(", null, i) - 2, x: ot(l, f) })) : f && g.push(r(g) + "rotate(" + f + i);
  }
  function a(l, f, g, v) {
    l !== f ? v.push({ i: g.push(r(g) + "skewX(", null, i) - 2, x: ot(l, f) }) : f && g.push(r(g) + "skewX(" + f + i);
  }
  function u(l, f, g, v, w, k) {
    if (l !== g || f !== v) {
      var b = w.push(r(w) + "scale(", null, ",", null, ")");
      k.push({ i: b - 4, x: ot(l, g) }, { i: b - 2, x: ot(f, v) });
    } else
      (g !== 1 || v !== 1) && w.push(r(w) + "scale(" + g + "," + v + ")");
  }
  return function(l, f) {
    var g = [], v = [];
    return l = t(l), f = t(f), o(l.translateX, l.translateY, f.translateX, f.translateY, g, v), s(l.rotate, f.rotate, g, v), a(l.skewX, f.skewX, g, v), u(l.scaleX, l.scaleY, f.scaleX, f.scaleY, g, v), l = f = null, function(w) {
      for (var k = -1, b = v.length, N; ++k < b; )
        g[(N = v[k]).i] = N.x(w);
      return g.join("");
    };
  };
}
var go = on(ho, "px, ", "px)", "deg)"), vo = on(mo, ", ", ")", ")"), po = 1e-12;
function Re(t) {
  return ((t = Math.exp(t)) + 1 / t) / 2;
}
function wo(t) {
  return ((t = Math.exp(t)) - 1 / t) / 2;
}
function yo(t) {
  return ((t = Math.exp(2 * t)) - 1) / (t + 1);
}
const _o = function t(e, n, i) {
  function r(o, s) {
    var a = o[0], u = o[1], l = o[2], f = s[0], g = s[1], v = s[2], w = f - a, k = g - u, b = w * w + k * k, N, x;
    if (b < po)
      x = Math.log(v / l) / e, N = function(V) {
        return [
          a + V * w,
          u + V * k,
          l * Math.exp(e * V * x)
        ];
      };
    else {
      var T = Math.sqrt(b), R = (v * v - l * l + i * b) / (2 * l * n * T), L = (v * v - l * l - i * b) / (2 * v * n * T), O = Math.log(Math.sqrt(R * R + 1) - R), A = Math.log(Math.sqrt(L * L + 1) - L);
      x = (A - O) / e, N = function(V) {
        var U = V * x, E = Re(O), ft = l / (n * T) * (E * yo(e * U + O) - wo(O));
        return [
          a + ft * w,
          u + ft * k,
          l * E / Re(e * U + O)
        ];
      };
    }
    return N.duration = x * 1e3 * e / Math.SQRT2, N;
  }
  return r.rho = function(o) {
    var s = Math.max(1e-3, +o), a = s * s, u = a * a;
    return t(s, a, u);
  }, r;
}(Math.SQRT2, 2, 4);
var gt = 0, _t = 0, wt = 0, sn = 1e3, Xt, xt, qt = 0, ct = 0, Gt = 0, Mt = typeof performance == "object" && performance.now ? performance : Date, an = typeof window == "object" && window.requestAnimationFrame ? window.requestAnimationFrame.bind(window) : function(t) {
  setTimeout(t, 17);
};
function pe() {
  return ct || (an(xo), ct = Mt.now() + Gt);
}
function xo() {
  ct = 0;
}
function Lt() {
  this._call = this._time = this._next = null;
}
Lt.prototype = un.prototype = {
  constructor: Lt,
  restart: function(t, e, n) {
    if (typeof t != "function")
      throw new TypeError("callback is not a function");
    n = (n == null ? pe() : +n) + (e == null ? 0 : +e), !this._next && xt !== this && (xt ? xt._next = this : Xt = this, xt = this), this._call = t, this._time = n, le();
  },
  stop: function() {
    this._call && (this._call = null, this._time = 1 / 0, le());
  }
};
function un(t, e, n) {
  var i = new Lt();
  return i.restart(t, e, n), i;
}
function bo() {
  pe(), ++gt;
  for (var t = Xt, e; t; )
    (e = ct - t._time) >= 0 && t._call.call(void 0, e), t = t._next;
  --gt;
}
function Ie() {
  ct = (qt = Mt.now()) + Gt, gt = _t = 0;
  try {
    bo();
  } finally {
    gt = 0, $o(), ct = 0;
  }
}
function ko() {
  var t = Mt.now(), e = t - qt;
  e > sn && (Gt -= e, qt = t);
}
function $o() {
  for (var t, e = Xt, n, i = 1 / 0; e; )
    e._call ? (i > e._time && (i = e._time), t = e, e = e._next) : (n = e._next, e._next = null, e = t ? t._next = n : Xt = n);
  xt = t, le(i);
}
function le(t) {
  if (!gt) {
    _t && (_t = clearTimeout(_t));
    var e = t - ct;
    e > 24 ? (t < 1 / 0 && (_t = setTimeout(Ie, t - Mt.now() - Gt)), wt && (wt = clearInterval(wt))) : (wt || (qt = Mt.now(), wt = setInterval(ko, sn)), gt = 1, an(Ie));
  }
}
function He(t, e, n) {
  var i = new Lt();
  return e = e == null ? 0 : +e, i.restart((r) => {
    i.stop(), t(r + e);
  }, e, n), i;
}
var Mo = de("start", "end", "cancel", "interrupt"), zo = [], ln = 0, Ve = 1, ce = 2, Ht = 3, De = 4, fe = 5, Vt = 6;
function Wt(t, e, n, i, r, o) {
  var s = t.__transition;
  if (!s)
    t.__transition = {};
  else if (n in s)
    return;
  Co(t, n, {
    name: e,
    index: i,
    group: r,
    on: Mo,
    tween: zo,
    time: o.time,
    delay: o.delay,
    duration: o.duration,
    ease: o.ease,
    timer: null,
    state: ln
  });
}
function we(t, e) {
  var n = W(t, e);
  if (n.state > ln)
    throw new Error("too late; already scheduled");
  return n;
}
function Q(t, e) {
  var n = W(t, e);
  if (n.state > Ht)
    throw new Error("too late; already running");
  return n;
}
function W(t, e) {
  var n = t.__transition;
  if (!n || !(n = n[e]))
    throw new Error("transition not found");
  return n;
}
function Co(t, e, n) {
  var i = t.__transition, r;
  i[e] = n, n.timer = un(o, 0, n.time);
  function o(l) {
    n.state = Ve, n.timer.restart(s, n.delay, n.time), n.delay <= l && s(l - n.delay);
  }
  function s(l) {
    var f, g, v, w;
    if (n.state !== Ve)
      return u();
    for (f in i)
      if (w = i[f], w.name === n.name) {
        if (w.state === Ht)
          return He(s);
        w.state === De ? (w.state = Vt, w.timer.stop(), w.on.call("interrupt", t, t.__data__, w.index, w.group), delete i[f]) : +f < e && (w.state = Vt, w.timer.stop(), w.on.call("cancel", t, t.__data__, w.index, w.group), delete i[f]);
      }
    if (He(function() {
      n.state === Ht && (n.state = De, n.timer.restart(a, n.delay, n.time), a(l));
    }), n.state = ce, n.on.call("start", t, t.__data__, n.index, n.group), n.state === ce) {
      for (n.state = Ht, r = new Array(v = n.tween.length), f = 0, g = -1; f < v; ++f)
        (w = n.tween[f].value.call(t, t.__data__, n.index, n.group)) && (r[++g] = w);
      r.length = g + 1;
    }
  }
  function a(l) {
    for (var f = l < n.duration ? n.ease.call(null, l / n.duration) : (n.timer.restart(u), n.state = fe, 1), g = -1, v = r.length; ++g < v; )
      r[g].call(t, f);
    n.state === fe && (n.on.call("end", t, t.__data__, n.index, n.group), u());
  }
  function u() {
    n.state = Vt, n.timer.stop(), delete i[e];
    for (var l in i)
      return;
    delete t.__transition;
  }
}
function Dt(t, e) {
  var n = t.__transition, i, r, o = !0, s;
  if (!!n) {
    e = e == null ? null : e + "";
    for (s in n) {
      if ((i = n[s]).name !== e) {
        o = !1;
        continue;
      }
      r = i.state > ce && i.state < fe, i.state = Vt, i.timer.stop(), i.on.call(r ? "interrupt" : "cancel", t, t.__data__, i.index, i.group), delete n[s];
    }
    o && delete t.__transition;
  }
}
function No(t) {
  return this.each(function() {
    Dt(this, t);
  });
}
function Eo(t, e) {
  var n, i;
  return function() {
    var r = Q(this, t), o = r.tween;
    if (o !== n) {
      i = n = o;
      for (var s = 0, a = i.length; s < a; ++s)
        if (i[s].name === e) {
          i = i.slice(), i.splice(s, 1);
          break;
        }
    }
    r.tween = i;
  };
}
function So(t, e, n) {
  var i, r;
  if (typeof n != "function")
    throw new Error();
  return function() {
    var o = Q(this, t), s = o.tween;
    if (s !== i) {
      r = (i = s).slice();
      for (var a = { name: e, value: n }, u = 0, l = r.length; u < l; ++u)
        if (r[u].name === e) {
          r[u] = a;
          break;
        }
      u === l && r.push(a);
    }
    o.tween = r;
  };
}
function Ao(t, e) {
  var n = this._id;
  if (t += "", arguments.length < 2) {
    for (var i = W(this.node(), n).tween, r = 0, o = i.length, s; r < o; ++r)
      if ((s = i[r]).name === t)
        return s.value;
    return null;
  }
  return this.each((e == null ? Eo : So)(n, t, e));
}
function ye(t, e, n) {
  var i = t._id;
  return t.each(function() {
    var r = Q(this, i);
    (r.value || (r.value = {}))[e] = n.apply(this, arguments);
  }), function(r) {
    return W(r, i).value[e];
  };
}
function cn(t, e) {
  var n;
  return (typeof e == "number" ? ot : e instanceof $t ? Ae : (n = $t(e)) ? (e = n, Ae) : fo)(t, e);
}
function To(t) {
  return function() {
    this.removeAttribute(t);
  };
}
function Ro(t) {
  return function() {
    this.removeAttributeNS(t.space, t.local);
  };
}
function Io(t, e, n) {
  var i, r = n + "", o;
  return function() {
    var s = this.getAttribute(t);
    return s === r ? null : s === i ? o : o = e(i = s, n);
  };
}
function Ho(t, e, n) {
  var i, r = n + "", o;
  return function() {
    var s = this.getAttributeNS(t.space, t.local);
    return s === r ? null : s === i ? o : o = e(i = s, n);
  };
}
function Vo(t, e, n) {
  var i, r, o;
  return function() {
    var s, a = n(this), u;
    return a == null ? void this.removeAttribute(t) : (s = this.getAttribute(t), u = a + "", s === u ? null : s === i && u === r ? o : (r = u, o = e(i = s, a)));
  };
}
function Do(t, e, n) {
  var i, r, o;
  return function() {
    var s, a = n(this), u;
    return a == null ? void this.removeAttributeNS(t.space, t.local) : (s = this.getAttributeNS(t.space, t.local), u = a + "", s === u ? null : s === i && u === r ? o : (r = u, o = e(i = s, a)));
  };
}
function Bo(t, e) {
  var n = Yt(t), i = n === "transform" ? vo : cn;
  return this.attrTween(t, typeof e == "function" ? (n.local ? Do : Vo)(n, i, ye(this, "attr." + t, e)) : e == null ? (n.local ? Ro : To)(n) : (n.local ? Ho : Io)(n, i, e));
}
function Fo(t, e) {
  return function(n) {
    this.setAttribute(t, e.call(this, n));
  };
}
function Oo(t, e) {
  return function(n) {
    this.setAttributeNS(t.space, t.local, e.call(this, n));
  };
}
function Xo(t, e) {
  var n, i;
  function r() {
    var o = e.apply(this, arguments);
    return o !== i && (n = (i = o) && Oo(t, o)), n;
  }
  return r._value = e, r;
}
function qo(t, e) {
  var n, i;
  function r() {
    var o = e.apply(this, arguments);
    return o !== i && (n = (i = o) && Fo(t, o)), n;
  }
  return r._value = e, r;
}
function Lo(t, e) {
  var n = "attr." + t;
  if (arguments.length < 2)
    return (n = this.tween(n)) && n._value;
  if (e == null)
    return this.tween(n, null);
  if (typeof e != "function")
    throw new Error();
  var i = Yt(t);
  return this.tween(n, (i.local ? Xo : qo)(i, e));
}
function Po(t, e) {
  return function() {
    we(this, t).delay = +e.apply(this, arguments);
  };
}
function Yo(t, e) {
  return e = +e, function() {
    we(this, t).delay = e;
  };
}
function Go(t) {
  var e = this._id;
  return arguments.length ? this.each((typeof t == "function" ? Po : Yo)(e, t)) : W(this.node(), e).delay;
}
function Wo(t, e) {
  return function() {
    Q(this, t).duration = +e.apply(this, arguments);
  };
}
function Uo(t, e) {
  return e = +e, function() {
    Q(this, t).duration = e;
  };
}
function Zo(t) {
  var e = this._id;
  return arguments.length ? this.each((typeof t == "function" ? Wo : Uo)(e, t)) : W(this.node(), e).duration;
}
function Ko(t, e) {
  if (typeof e != "function")
    throw new Error();
  return function() {
    Q(this, t).ease = e;
  };
}
function Qo(t) {
  var e = this._id;
  return arguments.length ? this.each(Ko(e, t)) : W(this.node(), e).ease;
}
function Jo(t, e) {
  return function() {
    var n = e.apply(this, arguments);
    if (typeof n != "function")
      throw new Error();
    Q(this, t).ease = n;
  };
}
function jo(t) {
  if (typeof t != "function")
    throw new Error();
  return this.each(Jo(this._id, t));
}
function ts(t) {
  typeof t != "function" && (t = Le(t));
  for (var e = this._groups, n = e.length, i = new Array(n), r = 0; r < n; ++r)
    for (var o = e[r], s = o.length, a = i[r] = [], u, l = 0; l < s; ++l)
      (u = o[l]) && t.call(u, u.__data__, l, o) && a.push(u);
  return new nt(i, this._parents, this._name, this._id);
}
function es(t) {
  if (t._id !== this._id)
    throw new Error();
  for (var e = this._groups, n = t._groups, i = e.length, r = n.length, o = Math.min(i, r), s = new Array(i), a = 0; a < o; ++a)
    for (var u = e[a], l = n[a], f = u.length, g = s[a] = new Array(f), v, w = 0; w < f; ++w)
      (v = u[w] || l[w]) && (g[w] = v);
  for (; a < i; ++a)
    s[a] = e[a];
  return new nt(s, this._parents, this._name, this._id);
}
function ns(t) {
  return (t + "").trim().split(/^|\s+/).every(function(e) {
    var n = e.indexOf(".");
    return n >= 0 && (e = e.slice(0, n)), !e || e === "start";
  });
}
function is(t, e, n) {
  var i, r, o = ns(e) ? we : Q;
  return function() {
    var s = o(this, t), a = s.on;
    a !== i && (r = (i = a).copy()).on(e, n), s.on = r;
  };
}
function rs(t, e) {
  var n = this._id;
  return arguments.length < 2 ? W(this.node(), n).on.on(t) : this.each(is(n, t, e));
}
function os(t) {
  return function() {
    var e = this.parentNode;
    for (var n in this.__transition)
      if (+n !== t)
        return;
    e && e.removeChild(this);
  };
}
function ss() {
  return this.on("end.remove", os(this._id));
}
function as(t) {
  var e = this._name, n = this._id;
  typeof t != "function" && (t = me(t));
  for (var i = this._groups, r = i.length, o = new Array(r), s = 0; s < r; ++s)
    for (var a = i[s], u = a.length, l = o[s] = new Array(u), f, g, v = 0; v < u; ++v)
      (f = a[v]) && (g = t.call(f, f.__data__, v, a)) && ("__data__" in f && (g.__data__ = f.__data__), l[v] = g, Wt(l[v], e, n, v, l, W(f, n)));
  return new nt(o, this._parents, e, n);
}
function us(t) {
  var e = this._name, n = this._id;
  typeof t != "function" && (t = qe(t));
  for (var i = this._groups, r = i.length, o = [], s = [], a = 0; a < r; ++a)
    for (var u = i[a], l = u.length, f, g = 0; g < l; ++g)
      if (f = u[g]) {
        for (var v = t.call(f, f.__data__, g, u), w, k = W(f, n), b = 0, N = v.length; b < N; ++b)
          (w = v[b]) && Wt(w, e, n, b, v, k);
        o.push(v), s.push(f);
      }
  return new nt(o, s, e, n);
}
var ls = Ct.prototype.constructor;
function cs() {
  return new ls(this._groups, this._parents);
}
function fs(t, e) {
  var n, i, r;
  return function() {
    var o = mt(this, t), s = (this.style.removeProperty(t), mt(this, t));
    return o === s ? null : o === n && s === i ? r : r = e(n = o, i = s);
  };
}
function fn(t) {
  return function() {
    this.style.removeProperty(t);
  };
}
function hs(t, e, n) {
  var i, r = n + "", o;
  return function() {
    var s = mt(this, t);
    return s === r ? null : s === i ? o : o = e(i = s, n);
  };
}
function ds(t, e, n) {
  var i, r, o;
  return function() {
    var s = mt(this, t), a = n(this), u = a + "";
    return a == null && (u = a = (this.style.removeProperty(t), mt(this, t))), s === u ? null : s === i && u === r ? o : (r = u, o = e(i = s, a));
  };
}
function ms(t, e) {
  var n, i, r, o = "style." + e, s = "end." + o, a;
  return function() {
    var u = Q(this, t), l = u.on, f = u.value[o] == null ? a || (a = fn(e)) : void 0;
    (l !== n || r !== f) && (i = (n = l).copy()).on(s, r = f), u.on = i;
  };
}
function gs(t, e, n) {
  var i = (t += "") == "transform" ? go : cn;
  return e == null ? this.styleTween(t, fs(t, i)).on("end.style." + t, fn(t)) : typeof e == "function" ? this.styleTween(t, ds(t, i, ye(this, "style." + t, e))).each(ms(this._id, t)) : this.styleTween(t, hs(t, i, e), n).on("end.style." + t, null);
}
function vs(t, e, n) {
  return function(i) {
    this.style.setProperty(t, e.call(this, i), n);
  };
}
function ps(t, e, n) {
  var i, r;
  function o() {
    var s = e.apply(this, arguments);
    return s !== r && (i = (r = s) && vs(t, s, n)), i;
  }
  return o._value = e, o;
}
function ws(t, e, n) {
  var i = "style." + (t += "");
  if (arguments.length < 2)
    return (i = this.tween(i)) && i._value;
  if (e == null)
    return this.tween(i, null);
  if (typeof e != "function")
    throw new Error();
  return this.tween(i, ps(t, e, n == null ? "" : n));
}
function ys(t) {
  return function() {
    this.textContent = t;
  };
}
function _s(t) {
  return function() {
    var e = t(this);
    this.textContent = e == null ? "" : e;
  };
}
function xs(t) {
  return this.tween("text", typeof t == "function" ? _s(ye(this, "text", t)) : ys(t == null ? "" : t + ""));
}
function bs(t) {
  return function(e) {
    this.textContent = t.call(this, e);
  };
}
function ks(t) {
  var e, n;
  function i() {
    var r = t.apply(this, arguments);
    return r !== n && (e = (n = r) && bs(r)), e;
  }
  return i._value = t, i;
}
function $s(t) {
  var e = "text";
  if (arguments.length < 1)
    return (e = this.tween(e)) && e._value;
  if (t == null)
    return this.tween(e, null);
  if (typeof t != "function")
    throw new Error();
  return this.tween(e, ks(t));
}
function Ms() {
  for (var t = this._name, e = this._id, n = hn(), i = this._groups, r = i.length, o = 0; o < r; ++o)
    for (var s = i[o], a = s.length, u, l = 0; l < a; ++l)
      if (u = s[l]) {
        var f = W(u, e);
        Wt(u, t, n, l, s, {
          time: f.time + f.delay + f.duration,
          delay: 0,
          duration: f.duration,
          ease: f.ease
        });
      }
  return new nt(i, this._parents, t, n);
}
function zs() {
  var t, e, n = this, i = n._id, r = n.size();
  return new Promise(function(o, s) {
    var a = { value: s }, u = { value: function() {
      --r === 0 && o();
    } };
    n.each(function() {
      var l = Q(this, i), f = l.on;
      f !== t && (e = (t = f).copy(), e._.cancel.push(a), e._.interrupt.push(a), e._.end.push(u)), l.on = e;
    }), r === 0 && o();
  });
}
var Cs = 0;
function nt(t, e, n, i) {
  this._groups = t, this._parents = e, this._name = n, this._id = i;
}
function hn() {
  return ++Cs;
}
var j = Ct.prototype;
nt.prototype = {
  constructor: nt,
  select: as,
  selectAll: us,
  selectChild: j.selectChild,
  selectChildren: j.selectChildren,
  filter: ts,
  merge: es,
  selection: cs,
  transition: Ms,
  call: j.call,
  nodes: j.nodes,
  node: j.node,
  size: j.size,
  empty: j.empty,
  each: j.each,
  on: rs,
  attr: Bo,
  attrTween: Lo,
  style: gs,
  styleTween: ws,
  text: xs,
  textTween: $s,
  remove: ss,
  tween: Ao,
  delay: Go,
  duration: Zo,
  ease: Qo,
  easeVarying: jo,
  end: zs,
  [Symbol.iterator]: j[Symbol.iterator]
};
function Ns(t) {
  return ((t *= 2) <= 1 ? t * t * t : (t -= 2) * t * t + 2) / 2;
}
var Es = {
  time: null,
  delay: 0,
  duration: 250,
  ease: Ns
};
function Ss(t, e) {
  for (var n; !(n = t.__transition) || !(n = n[e]); )
    if (!(t = t.parentNode))
      throw new Error(`transition ${e} not found`);
  return n;
}
function As(t) {
  var e, n;
  t instanceof nt ? (e = t._id, t = t._name) : (e = hn(), (n = Es).time = pe(), t = t == null ? null : t + "");
  for (var i = this._groups, r = i.length, o = 0; o < r; ++o)
    for (var s = i[o], a = s.length, u, l = 0; l < a; ++l)
      (u = s[l]) && Wt(u, t, e, l, s, n || Ss(u, e));
  return new nt(i, this._parents, t, e);
}
Ct.prototype.interrupt = No;
Ct.prototype.transition = As;
const Rt = (t) => () => t;
function Ts(t, {
  sourceEvent: e,
  target: n,
  transform: i,
  dispatch: r
}) {
  Object.defineProperties(this, {
    type: { value: t, enumerable: !0, configurable: !0 },
    sourceEvent: { value: e, enumerable: !0, configurable: !0 },
    target: { value: n, enumerable: !0, configurable: !0 },
    transform: { value: i, enumerable: !0, configurable: !0 },
    _: { value: r }
  });
}
function et(t, e, n) {
  this.k = t, this.x = e, this.y = n;
}
et.prototype = {
  constructor: et,
  scale: function(t) {
    return t === 1 ? this : new et(this.k * t, this.x, this.y);
  },
  translate: function(t, e) {
    return t === 0 & e === 0 ? this : new et(this.k, this.x + this.k * t, this.y + this.k * e);
  },
  apply: function(t) {
    return [t[0] * this.k + this.x, t[1] * this.k + this.y];
  },
  applyX: function(t) {
    return t * this.k + this.x;
  },
  applyY: function(t) {
    return t * this.k + this.y;
  },
  invert: function(t) {
    return [(t[0] - this.x) / this.k, (t[1] - this.y) / this.k];
  },
  invertX: function(t) {
    return (t - this.x) / this.k;
  },
  invertY: function(t) {
    return (t - this.y) / this.k;
  },
  rescaleX: function(t) {
    return t.copy().domain(t.range().map(this.invertX, this).map(t.invert, t));
  },
  rescaleY: function(t) {
    return t.copy().domain(t.range().map(this.invertY, this).map(t.invert, t));
  },
  toString: function() {
    return "translate(" + this.x + "," + this.y + ") scale(" + this.k + ")";
  }
};
var _e = new et(1, 0, 0);
et.prototype;
function ee(t) {
  t.stopImmediatePropagation();
}
function yt(t) {
  t.preventDefault(), t.stopImmediatePropagation();
}
function Rs(t) {
  return (!t.ctrlKey || t.type === "wheel") && !t.button;
}
function Is() {
  var t = this;
  return t instanceof SVGElement ? (t = t.ownerSVGElement || t, t.hasAttribute("viewBox") ? (t = t.viewBox.baseVal, [[t.x, t.y], [t.x + t.width, t.y + t.height]]) : [[0, 0], [t.width.baseVal.value, t.height.baseVal.value]]) : [[0, 0], [t.clientWidth, t.clientHeight]];
}
function Be() {
  return this.__zoom || _e;
}
function Hs(t) {
  return -t.deltaY * (t.deltaMode === 1 ? 0.05 : t.deltaMode ? 1 : 2e-3) * (t.ctrlKey ? 10 : 1);
}
function Vs() {
  return navigator.maxTouchPoints || "ontouchstart" in this;
}
function Ds(t, e, n) {
  var i = t.invertX(e[0][0]) - n[0][0], r = t.invertX(e[1][0]) - n[1][0], o = t.invertY(e[0][1]) - n[0][1], s = t.invertY(e[1][1]) - n[1][1];
  return t.translate(
    r > i ? (i + r) / 2 : Math.min(0, i) || Math.max(0, r),
    s > o ? (o + s) / 2 : Math.min(0, o) || Math.max(0, s)
  );
}
function Bs() {
  var t = Rs, e = Is, n = Ds, i = Hs, r = Vs, o = [0, 1 / 0], s = [[-1 / 0, -1 / 0], [1 / 0, 1 / 0]], a = 250, u = _o, l = de("start", "zoom", "end"), f, g, v, w = 500, k = 150, b = 0, N = 10;
  function x(c) {
    c.property("__zoom", Be).on("wheel.zoom", U, { passive: !1 }).on("mousedown.zoom", E).on("dblclick.zoom", ft).filter(r).on("touchstart.zoom", Ut).on("touchmove.zoom", Zt).on("touchend.zoom touchcancel.zoom", Kt).style("-webkit-tap-highlight-color", "rgba(0,0,0,0)");
  }
  x.transform = function(c, y, h, m) {
    var p = c.selection ? c.selection() : c;
    p.property("__zoom", Be), c !== p ? O(c, y, h, m) : p.interrupt().each(function() {
      A(this, arguments).event(m).start().zoom(null, typeof y == "function" ? y.apply(this, arguments) : y).end();
    });
  }, x.scaleBy = function(c, y, h, m) {
    x.scaleTo(c, function() {
      var p = this.__zoom.k, d = typeof y == "function" ? y.apply(this, arguments) : y;
      return p * d;
    }, h, m);
  }, x.scaleTo = function(c, y, h, m) {
    x.transform(c, function() {
      var p = e.apply(this, arguments), d = this.__zoom, _ = h == null ? L(p) : typeof h == "function" ? h.apply(this, arguments) : h, z = d.invert(_), $ = typeof y == "function" ? y.apply(this, arguments) : y;
      return n(R(T(d, $), _, z), p, s);
    }, h, m);
  }, x.translateBy = function(c, y, h, m) {
    x.transform(c, function() {
      return n(this.__zoom.translate(
        typeof y == "function" ? y.apply(this, arguments) : y,
        typeof h == "function" ? h.apply(this, arguments) : h
      ), e.apply(this, arguments), s);
    }, null, m);
  }, x.translateTo = function(c, y, h, m, p) {
    x.transform(c, function() {
      var d = e.apply(this, arguments), _ = this.__zoom, z = m == null ? L(d) : typeof m == "function" ? m.apply(this, arguments) : m;
      return n(_e.translate(z[0], z[1]).scale(_.k).translate(
        typeof y == "function" ? -y.apply(this, arguments) : -y,
        typeof h == "function" ? -h.apply(this, arguments) : -h
      ), d, s);
    }, m, p);
  };
  function T(c, y) {
    return y = Math.max(o[0], Math.min(o[1], y)), y === c.k ? c : new et(y, c.x, c.y);
  }
  function R(c, y, h) {
    var m = y[0] - h[0] * c.k, p = y[1] - h[1] * c.k;
    return m === c.x && p === c.y ? c : new et(c.k, m, p);
  }
  function L(c) {
    return [(+c[0][0] + +c[1][0]) / 2, (+c[0][1] + +c[1][1]) / 2];
  }
  function O(c, y, h, m) {
    c.on("start.zoom", function() {
      A(this, arguments).event(m).start();
    }).on("interrupt.zoom end.zoom", function() {
      A(this, arguments).event(m).end();
    }).tween("zoom", function() {
      var p = this, d = arguments, _ = A(p, d).event(m), z = e.apply(p, d), $ = h == null ? L(z) : typeof h == "function" ? h.apply(p, d) : h, D = Math.max(z[1][0] - z[0][0], z[1][1] - z[0][1]), S = p.__zoom, X = typeof y == "function" ? y.apply(p, d) : y, P = u(S.invert($).concat(D / S.k), X.invert($).concat(D / X.k));
      return function(Y) {
        if (Y === 1)
          Y = X;
        else {
          var J = P(Y), Qt = D / J[2];
          Y = new et(Qt, $[0] - J[0] * Qt, $[1] - J[1] * Qt);
        }
        _.zoom(null, Y);
      };
    });
  }
  function A(c, y, h) {
    return !h && c.__zooming || new V(c, y);
  }
  function V(c, y) {
    this.that = c, this.args = y, this.active = 0, this.sourceEvent = null, this.extent = e.apply(c, y), this.taps = 0;
  }
  V.prototype = {
    event: function(c) {
      return c && (this.sourceEvent = c), this;
    },
    start: function() {
      return ++this.active === 1 && (this.that.__zooming = this, this.emit("start")), this;
    },
    zoom: function(c, y) {
      return this.mouse && c !== "mouse" && (this.mouse[1] = y.invert(this.mouse[0])), this.touch0 && c !== "touch" && (this.touch0[1] = y.invert(this.touch0[0])), this.touch1 && c !== "touch" && (this.touch1[1] = y.invert(this.touch1[0])), this.that.__zoom = y, this.emit("zoom"), this;
    },
    end: function() {
      return --this.active === 0 && (delete this.that.__zooming, this.emit("end")), this;
    },
    emit: function(c) {
      var y = st(this.that).datum();
      l.call(
        c,
        this.that,
        new Ts(c, {
          sourceEvent: this.sourceEvent,
          target: x,
          type: c,
          transform: this.that.__zoom,
          dispatch: l
        }),
        y
      );
    }
  };
  function U(c, ...y) {
    if (!t.apply(this, arguments))
      return;
    var h = A(this, y).event(c), m = this.__zoom, p = Math.max(o[0], Math.min(o[1], m.k * Math.pow(2, i.apply(this, arguments)))), d = it(c);
    if (h.wheel)
      (h.mouse[0][0] !== d[0] || h.mouse[0][1] !== d[1]) && (h.mouse[1] = m.invert(h.mouse[0] = d)), clearTimeout(h.wheel);
    else {
      if (m.k === p)
        return;
      h.mouse = [d, m.invert(d)], Dt(this), h.start();
    }
    yt(c), h.wheel = setTimeout(_, k), h.zoom("mouse", n(R(T(m, p), h.mouse[0], h.mouse[1]), h.extent, s));
    function _() {
      h.wheel = null, h.end();
    }
  }
  function E(c, ...y) {
    if (v || !t.apply(this, arguments))
      return;
    var h = c.currentTarget, m = A(this, y, !0).event(c), p = st(c.view).on("mousemove.zoom", $, !0).on("mouseup.zoom", D, !0), d = it(c, h), _ = c.clientX, z = c.clientY;
    Gr(c.view), ee(c), m.mouse = [d, this.__zoom.invert(d)], Dt(this), m.start();
    function $(S) {
      if (yt(S), !m.moved) {
        var X = S.clientX - _, P = S.clientY - z;
        m.moved = X * X + P * P > b;
      }
      m.event(S).zoom("mouse", n(R(m.that.__zoom, m.mouse[0] = it(S, h), m.mouse[1]), m.extent, s));
    }
    function D(S) {
      p.on("mousemove.zoom mouseup.zoom", null), Wr(S.view, m.moved), yt(S), m.event(S).end();
    }
  }
  function ft(c, ...y) {
    if (!!t.apply(this, arguments)) {
      var h = this.__zoom, m = it(c.changedTouches ? c.changedTouches[0] : c, this), p = h.invert(m), d = h.k * (c.shiftKey ? 0.5 : 2), _ = n(R(T(h, d), m, p), e.apply(this, y), s);
      yt(c), a > 0 ? st(this).transition().duration(a).call(O, _, m, c) : st(this).call(x.transform, _, m, c);
    }
  }
  function Ut(c, ...y) {
    if (!!t.apply(this, arguments)) {
      var h = c.touches, m = h.length, p = A(this, y, c.changedTouches.length === m).event(c), d, _, z, $;
      for (ee(c), _ = 0; _ < m; ++_)
        z = h[_], $ = it(z, this), $ = [$, this.__zoom.invert($), z.identifier], p.touch0 ? !p.touch1 && p.touch0[2] !== $[2] && (p.touch1 = $, p.taps = 0) : (p.touch0 = $, d = !0, p.taps = 1 + !!f);
      f && (f = clearTimeout(f)), d && (p.taps < 2 && (g = $[0], f = setTimeout(function() {
        f = null;
      }, w)), Dt(this), p.start());
    }
  }
  function Zt(c, ...y) {
    if (!!this.__zooming) {
      var h = A(this, y).event(c), m = c.changedTouches, p = m.length, d, _, z, $;
      for (yt(c), d = 0; d < p; ++d)
        _ = m[d], z = it(_, this), h.touch0 && h.touch0[2] === _.identifier ? h.touch0[0] = z : h.touch1 && h.touch1[2] === _.identifier && (h.touch1[0] = z);
      if (_ = h.that.__zoom, h.touch1) {
        var D = h.touch0[0], S = h.touch0[1], X = h.touch1[0], P = h.touch1[1], Y = (Y = X[0] - D[0]) * Y + (Y = X[1] - D[1]) * Y, J = (J = P[0] - S[0]) * J + (J = P[1] - S[1]) * J;
        _ = T(_, Math.sqrt(Y / J)), z = [(D[0] + X[0]) / 2, (D[1] + X[1]) / 2], $ = [(S[0] + P[0]) / 2, (S[1] + P[1]) / 2];
      } else if (h.touch0)
        z = h.touch0[0], $ = h.touch0[1];
      else
        return;
      h.zoom("touch", n(R(_, z, $), h.extent, s));
    }
  }
  function Kt(c, ...y) {
    if (!!this.__zooming) {
      var h = A(this, y).event(c), m = c.changedTouches, p = m.length, d, _;
      for (ee(c), v && clearTimeout(v), v = setTimeout(function() {
        v = null;
      }, w), d = 0; d < p; ++d)
        _ = m[d], h.touch0 && h.touch0[2] === _.identifier ? delete h.touch0 : h.touch1 && h.touch1[2] === _.identifier && delete h.touch1;
      if (h.touch1 && !h.touch0 && (h.touch0 = h.touch1, delete h.touch1), h.touch0)
        h.touch0[1] = this.__zoom.invert(h.touch0[0]);
      else if (h.end(), h.taps === 2 && (_ = it(_, this), Math.hypot(g[0] - _[0], g[1] - _[1]) < N)) {
        var z = st(this).on("dblclick.zoom");
        z && z.apply(this, arguments);
      }
    }
  }
  return x.wheelDelta = function(c) {
    return arguments.length ? (i = typeof c == "function" ? c : Rt(+c), x) : i;
  }, x.filter = function(c) {
    return arguments.length ? (t = typeof c == "function" ? c : Rt(!!c), x) : t;
  }, x.touchable = function(c) {
    return arguments.length ? (r = typeof c == "function" ? c : Rt(!!c), x) : r;
  }, x.extent = function(c) {
    return arguments.length ? (e = typeof c == "function" ? c : Rt([[+c[0][0], +c[0][1]], [+c[1][0], +c[1][1]]]), x) : e;
  }, x.scaleExtent = function(c) {
    return arguments.length ? (o[0] = +c[0], o[1] = +c[1], x) : [o[0], o[1]];
  }, x.translateExtent = function(c) {
    return arguments.length ? (s[0][0] = +c[0][0], s[1][0] = +c[1][0], s[0][1] = +c[0][1], s[1][1] = +c[1][1], x) : [[s[0][0], s[0][1]], [s[1][0], s[1][1]]];
  }, x.constrain = function(c) {
    return arguments.length ? (n = c, x) : n;
  }, x.duration = function(c) {
    return arguments.length ? (a = +c, x) : a;
  }, x.interpolate = function(c) {
    return arguments.length ? (u = c, x) : u;
  }, x.on = function() {
    var c = l.on.apply(l, arguments);
    return c === l ? x : c;
  }, x.clickDistance = function(c) {
    return arguments.length ? (b = (c = +c) * c, x) : Math.sqrt(b);
  }, x.tapDistance = function(c) {
    return arguments.length ? (N = +c, x) : N;
  }, x;
}
const dn = Symbol("MiniMapSlots"), Fs = zt({
  name: "MiniMapNode",
  props: [
    "id",
    "position",
    "dimensions",
    "strokeWidth",
    "strokeColor",
    "borderRadius",
    "color",
    "shapeRendering",
    "type",
    "selected",
    "dragging"
  ],
  emits: ["click", "dblclick", "mouseenter", "mousemove", "mouseleave"],
  setup(t, { attrs: e, emit: n }) {
    const i = mn(dn);
    return () => {
      var s;
      const r = (s = e.style) != null ? s : {}, o = i[`node-${t.type}`];
      return o ? o(t) : gn("rect", {
        id: t.id,
        class: ["vue-flow__minimap-node", e.class].join(" "),
        style: r,
        x: t.position.x,
        y: t.position.y,
        rx: t.borderRadius,
        ry: t.borderRadius,
        width: t.dimensions.width,
        height: t.dimensions.height,
        fill: t.color || r.background || r.backgroundColor,
        stroke: t.strokeColor,
        strokeWidth: t.strokeWidth,
        shapeRendering: t.shapeRendering,
        onClick: (a) => n("click", a),
        onDblClick: (a) => n("dblclick", a),
        onMouseenter: (a) => n("mouseenter", a),
        onMousemove: (a) => n("mousemove", a),
        onMouseleave: (a) => n("mouseleave", a)
      });
    };
  }
}), Os = ["width", "height", "viewBox", "aria-labelledby"], Xs = ["id"], qs = ["d", "fill", "stroke", "stroke-width"], Ls = {
  name: "MiniMap"
}, Us = /* @__PURE__ */ zt({
  ...Ls,
  props: {
    nodeColor: { type: [String, Function], default: "#e2e2e2" },
    nodeStrokeColor: { type: [String, Function], default: "transparent" },
    nodeClassName: null,
    nodeBorderRadius: { default: 5 },
    nodeStrokeWidth: { default: 2 },
    maskColor: { default: "rgb(240, 240, 240, 0.6)" },
    maskStrokeColor: { default: "none" },
    maskStrokeWidth: { default: 1 },
    position: { default: () => "bottom-right" },
    pannable: { type: Boolean, default: !1 },
    zoomable: { type: Boolean, default: !1 },
    width: null,
    height: null
  },
  emits: ["click", "nodeClick", "nodeDblclick", "nodeMouseenter", "nodeMousemove", "nodeMouseleave"],
  setup(t, { emit: e }) {
    const n = vn(), i = 200, r = 150, { id: o, edges: s, viewport: a, translateExtent: u, dimensions: l, emits: f, nodes: g, d3Selection: v, d3Zoom: w } = Pt(), k = pn();
    wn(dn, yn());
    const b = I(() => {
      var m, p, d;
      return (d = (p = t.width) != null ? p : (m = n.style) == null ? void 0 : m.width) != null ? d : i;
    }), N = I(() => {
      var m, p, d;
      return (d = (p = t.height) != null ? p : (m = n.style) == null ? void 0 : m.height) != null ? d : r;
    }), x = typeof window > "u" || !!window.chrome ? "crispEdges" : "geometricPrecision", T = I(() => t.nodeColor instanceof Function ? t.nodeColor : () => t.nodeColor), R = I(
      () => t.nodeStrokeColor instanceof Function ? t.nodeStrokeColor : () => t.nodeStrokeColor
    ), L = I(
      () => t.nodeClassName instanceof Function ? t.nodeClassName : () => t.nodeClassName
    ), O = I(() => kn(g.value)), A = I(() => ({
      x: -a.value.x / a.value.zoom,
      y: -a.value.y / a.value.zoom,
      width: l.value.width / a.value.zoom,
      height: l.value.height / a.value.zoom
    })), V = I(() => g.value && g.value.length ? $n(O.value, A.value) : A.value), U = I(() => {
      const m = V.value.width / b.value, p = V.value.height / N.value;
      return Math.max(m, p);
    }), E = I(() => {
      const m = U.value * b.value, p = U.value * N.value, d = 5 * U.value;
      return {
        offset: d,
        x: V.value.x - (m - V.value.width) / 2 - d,
        y: V.value.y - (p - V.value.height) / 2 - d,
        width: m + d * 2,
        height: p + d * 2
      };
    }), ft = I(() => !E.value.x || !E.value.y ? "" : `
    M${E.value.x - E.value.offset},${E.value.y - E.value.offset}
    h${E.value.width + E.value.offset * 2}
    v${E.value.height + E.value.offset * 2}
    h${-E.value.width - E.value.offset * 2}z
    M${A.value.x},${A.value.y}
    h${A.value.width}
    v${A.value.height}
    h${-A.value.width}z`);
    _n(
      (m) => {
        if (k.value) {
          const p = st(k.value), d = ($) => {
            if ($.sourceEvent.type !== "wheel" || !v.value || !w.value)
              return;
            const D = -$.sourceEvent.deltaY * ($.sourceEvent.deltaMode === 1 ? 0.05 : $.sourceEvent.deltaMode ? 1 : 2e-3) * 10, S = a.value.zoom * 2 ** D;
            w.value.scaleTo(v.value, S);
          }, _ = ($) => {
            if ($.sourceEvent.type !== "mousemove" || !v.value || !w.value)
              return;
            const D = {
              x: a.value.x - $.sourceEvent.movementX * U.value * Math.max(1, a.value.zoom),
              y: a.value.y - $.sourceEvent.movementY * U.value * Math.max(1, a.value.zoom)
            }, S = [
              [0, 0],
              [l.value.width, l.value.height]
            ], X = _e.translate(D.x, D.y).scale(a.value.zoom), P = w.value.constrain()(X, S, u.value);
            w.value.transform(v.value, P);
          }, z = Bs().on("zoom", t.pannable ? _ : () => {
          }).on("zoom.wheel", t.zoomable ? d : () => {
          });
          p.call(z), m(() => {
            p.on("zoom", null);
          });
        }
      },
      { flush: "post" }
    );
    const Ut = (m) => {
      const [p, d] = it(m);
      e("click", { event: m, position: { x: p, y: d } });
    }, Zt = (m, p) => {
      const d = { event: m, node: p, connectedEdges: pt([p], s.value) };
      f.miniMapNodeClick(d), e("nodeClick", d);
    }, Kt = (m, p) => {
      const d = { event: m, node: p, connectedEdges: pt([p], s.value) };
      f.miniMapNodeDoubleClick(d), e("nodeDblclick", d);
    }, c = (m, p) => {
      const d = { event: m, node: p, connectedEdges: pt([p], s.value) };
      f.miniMapNodeMouseEnter(d), e("nodeMouseenter", d);
    }, y = (m, p) => {
      const d = { event: m, node: p, connectedEdges: pt([p], s.value) };
      f.miniMapNodeMouseMove(d), e("nodeMousemove", d);
    }, h = (m, p) => {
      const d = { event: m, node: p, connectedEdges: pt([p], s.value) };
      f.miniMapNodeMouseLeave(d), e("nodeMouseleave", d);
    };
    return (m, p) => (C(), tt(M(Oe), {
      position: t.position,
      class: ne(["vue-flow__minimap", { pannable: t.pannable, zoomable: t.zoomable }])
    }, {
      default: ht(() => [
        (C(), B("svg", {
          ref_key: "el",
          ref: k,
          width: M(b),
          height: M(N),
          viewBox: [M(E).x, M(E).y, M(E).width, M(E).height].join(" "),
          role: "img",
          "aria-labelledby": `vue-flow__minimap-${M(o)}`,
          onClick: Ut
        }, [
          Z("title", {
            id: `vue-flow__minimap-${M(o)}`
          }, "Vue Flow mini map " + xn(M(o)), 9, Xs),
          (C(!0), B(Fe, null, bn(M(g), (d) => (C(), tt(M(Fs), {
            id: d.id,
            key: d.id,
            position: d.computedPosition,
            dimensions: d.dimensions,
            style: he(d.style),
            selected: d.selected,
            dragging: d.dragging,
            class: ne(M(L)(d)),
            color: M(T)(d),
            "border-radius": t.nodeBorderRadius,
            "stroke-color": M(R)(d),
            "stroke-width": t.nodeStrokeWidth,
            "shape-rendering": M(x),
            type: d.type,
            onClick: (_) => Zt(_, d),
            onDblclick: (_) => Kt(_, d),
            onMouseenter: (_) => c(_, d),
            onMousemove: (_) => y(_, d),
            onMouseleave: (_) => h(_, d)
          }, null, 8, ["id", "position", "dimensions", "style", "selected", "dragging", "class", "color", "border-radius", "stroke-color", "stroke-width", "shape-rendering", "type", "onClick", "onDblclick", "onMouseenter", "onMousemove", "onMouseleave"]))), 128)),
          Z("path", {
            class: "vue-flow__minimap-mask",
            d: M(ft),
            fill: t.maskColor,
            stroke: t.maskStrokeColor,
            "stroke-width": t.maskStrokeWidth,
            "fill-rule": "evenodd"
          }, null, 8, qs)
        ], 8, Os))
      ]),
      _: 1
    }, 8, ["position", "class"]));
  }
});
export {
  Gs as Background,
  Tn as BackgroundVariant,
  Et as ControlButton,
  Ws as Controls,
  Us as MiniMap,
  Fs as MiniMapNode,
  dn as MiniMapSlots,
  Oe as Panel,
  Rn as PanelPosition
};
